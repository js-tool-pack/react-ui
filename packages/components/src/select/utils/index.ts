export * from './pickOptionsByValue';
export * from './flattenOptions';
export * from './isValueChanged';
export * from './filterOptions';
export * from './isDivider';
export * from './intoView';
export * from './isGroup';
