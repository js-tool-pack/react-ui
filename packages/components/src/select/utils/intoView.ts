export function intoView(child: HTMLElement, parent: HTMLElement): void {
  // 底部往顶部走
  if (parent.scrollTop > child.offsetTop) {
    parent.scrollTop = child.offsetTop;
    return;
  }
  // 顶部往底部走
  if (
    parent.scrollTop + parent.offsetHeight <
    child.offsetTop + child.offsetHeight
  ) {
    parent.scrollTop =
      child.offsetTop - parent.offsetHeight + child.offsetHeight;
  }
}
