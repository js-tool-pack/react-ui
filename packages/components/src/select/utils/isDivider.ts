import type { SelectOptionsItem, SelectDivider } from '../select.types';

export function isDivider(opt: SelectOptionsItem): opt is SelectDivider {
  return (opt as SelectDivider).type === 'divider';
}
