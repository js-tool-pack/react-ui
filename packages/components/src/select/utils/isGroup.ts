import type { SelectOptionsItem, SelectOptionGroup } from '../select.types';

export function isGroup(opt: SelectOptionsItem): opt is SelectOptionGroup {
  return (opt as SelectOptionGroup).type === 'group';
}
