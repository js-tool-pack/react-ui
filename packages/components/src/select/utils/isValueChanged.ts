import { SelectOption } from '~/select';
import React from 'react';

export function isValueChanged(
  value: React.Key[] | React.Key | undefined,
  selectedOptions: SelectOption[],
): boolean {
  if (value === undefined) return false;
  if (!Array.isArray(value))
    return selectedOptions.length !== 1 || selectedOptions[0]?.value !== value;
  if (value.length !== selectedOptions.length) return true;
  return selectedOptions.some((v, i) => value[i] !== v.value);
}
