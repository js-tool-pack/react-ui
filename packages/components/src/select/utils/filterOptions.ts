import type { SelectOption, SelectOptionsItem } from '../select.types';
import { isString } from '@tool-pack/basic';
import { isDivider } from './isDivider';
import { isGroup } from './isGroup';

export function filterOptions(
  options: SelectOptionsItem[],
  pattern: string,
  filter = filterLabel,
): SelectOptionsItem[] {
  const result: SelectOptionsItem[] = [];
  options.forEach((opt) => {
    if (isDivider(opt)) return;
    if (isGroup(opt)) {
      if (!opt.children?.length) return;
      const children = filterOptions(opt.children, pattern, filter);
      if (!children.length) return;
      result.push({ ...opt, children });
      return;
    }
    if (filter(pattern, opt)) result.push(opt);
  });
  return result;
}

function filterLabel(pattern: string, option: SelectOption): boolean {
  return isString(option.label) && option.label.includes(pattern);
}
