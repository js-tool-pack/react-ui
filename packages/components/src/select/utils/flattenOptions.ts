import type { SelectOptionsItem, SelectOption } from '../select.types';
import { isDivider } from './isDivider';
import { isGroup } from './isGroup';

export function flattenOptions(
  options: SelectOptionsItem[],
  filter: (option: SelectOption) => boolean = () => true,
): SelectOption[] {
  const result: SelectOption[] = [];
  const queue = options.slice();
  while (queue.length) {
    const opt = queue.shift()!;
    if (isDivider(opt)) continue;
    if (isGroup(opt)) {
      queue.unshift(...(opt.children || []));
      continue;
    }
    if (!filter(opt)) continue;
    result.push(opt);
  }
  return result;
}
