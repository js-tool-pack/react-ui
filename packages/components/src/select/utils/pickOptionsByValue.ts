import { SelectOptionsItem, SelectOption } from '~/select';
import { flattenOptions } from '~/select/utils';
import React from 'react';

export function pickOptionsByValue(
  options: SelectOptionsItem[],
  value: React.Key[] | React.Key | undefined,
): SelectOption[] {
  if (value === undefined) return [];
  const filter: (option: SelectOption) => boolean = Array.isArray(value)
    ? (opt) => value.includes(opt.value)
    : (opt) => opt.value === value;
  return flattenOptions(options, filter);
}
