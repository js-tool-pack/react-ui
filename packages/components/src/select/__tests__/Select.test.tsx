import {
  SelectControllerRef,
  SelectOptionsItem,
  SelectOption,
  Select,
} from '~/select';
import { fireEvent, render, act } from '@testing-library/react';
import { getBalloon, $$, $ } from './utils';
import { emptyFn } from '@tool-pack/basic';
import { testAttrs } from '~/testAttrs';
import { useRef } from 'react';

describe('Select', () => {
  testAttrs(Select);

  test('basic', () => {
    const options: SelectOptionsItem[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
    ];
    const { container } = render(<Select options={options} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('select', () => {
    const options: SelectOption[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
    ];
    test('single', () => {
      const onSelect = jest.fn();
      const onChange = jest.fn();
      render(
        <Select
          onChange={onChange}
          onSelect={onSelect}
          options={options}
          visible
        />,
      );
      expect(onSelect).not.toBeCalled();
      expect(onChange).not.toBeCalled();
      expect(getBalloon()).not.toBeNull();

      fireEvent.click($('.t-select-option')!);
      expect(onSelect).toBeCalled();
      expect(onChange).toBeCalled();

      expect(onSelect.mock.calls[0]).toEqual([options[0]]);
      expect(onChange.mock.calls[0]).toEqual([options[0]!.value, [options[0]]]);
    });
    test('multiple', () => {
      const onSelect = jest.fn();
      const onChange = jest.fn();

      render(
        <Select
          onChange={onChange}
          onSelect={onSelect}
          options={options}
          multiple
          visible
        />,
      );
      expect(onSelect).not.toBeCalled();
      expect(onChange).not.toBeCalled();
      expect(getBalloon()).not.toBeNull();

      fireEvent.click($('.t-select-option')!);
      expect(onSelect).toBeCalled();
      expect(onChange).toBeCalled();

      expect(onSelect.mock.calls[0]).toEqual([options[0]]);
      expect(onChange.mock.calls[0]).toEqual([
        [options[0]!.value],
        [options[0]],
      ]);

      onSelect.mockClear();
      onChange.mockClear();

      fireEvent.click($$('.t-select-option')[1]!);
      expect(onSelect.mock.calls[0]).toEqual([options[1]]);
      expect(onChange.mock.calls[0]).toEqual([
        options.map((o) => o.value),
        options,
      ]);
    });
  });

  test('disabled', () => {
    const options: SelectOptionsItem[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
    ];
    const { container } = render(<Select options={options} disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('group', () => {
    const options: SelectOptionsItem[] = [
      {
        children: [
          {
            label: 'foo',
            value: 2,
          },
          {
            label: 'bar',
            value: 3,
          },
        ],
        attrs: { className: 'test' },
        type: 'group',
        label: '1',
        key: 1,
      },
      {
        label: 'foo',
        value: 4,
      },
      {
        label: 'bar',
        value: 5,
      },
    ];
    render(<Select options={options} visible={true} />);
    expect(document.body).toMatchSnapshot();
  });

  test('clearable|onClear', () => {
    const onClear = jest.fn();
    const options: SelectOptionsItem[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
    ];
    const { container } = render(
      <Select onClear={onClear} options={options} clearable value={1} />,
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(
      container.querySelector('.t-select-selection__selected')!.firstChild,
    ).toHaveClass('t-select-selection__label');
    expect(
      container.querySelector('.t-select-selection__selected')!.firstChild,
    ).not.toHaveClass('t-select-selection__placeholder');

    expect(onClear).not.toBeCalled();
    fireEvent.click(container.querySelector('.t-select-selection__clear')!);
    expect(onClear).toBeCalled();
    expect(
      container.querySelector('.t-select-selection__selected')!.firstChild,
    ).not.toHaveClass('t-select-selection__label');
    expect(
      container.querySelector('.t-select-selection__selected')!.firstChild,
    ).toHaveClass('t-select-selection__placeholder');
  });

  test('multiple', () => {
    const options: SelectOptionsItem[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
    ];
    render(<Select options={options} value={[1, 2]} multiple visible />);
    expect(document.body).toMatchSnapshot();
  });

  test('maxTagCount', () => {
    jest.useFakeTimers();
    const options: SelectOptionsItem[] = [
      {
        label: 'foo',
        value: 1,
      },
      {
        label: 'bar',
        value: 2,
      },
      {
        label: 'bar',
        value: 3,
      },
      {
        label: 'bar',
        value: 4,
      },
      {
        label: 'bar',
        value: 5,
      },
    ];
    render(
      <Select options={options} maxTagCount={1} value={[1, 2]} multiple />,
    );
    expect(document.body).toMatchSnapshot();

    fireEvent.mouseEnter($('.t-select-tags__count')!);
    act(() => jest.advanceTimersByTime(500));
    expect(document.body).toMatchSnapshot();
  });
  test('controller|onFocus|onBlur', () => {
    jest.useFakeTimers();
    const onFocus = jest.fn();
    const onBlur = jest.fn();

    render(<App />);

    expect(onFocus).not.toBeCalled();
    expect(onBlur).not.toBeCalled();

    expect(getTrigger()).not.toHaveFocus();
    fireEvent.click($('#focus')!);
    expect(getTrigger()).toHaveFocus();
    expect(onFocus).toBeCalled();
    expect(onBlur).not.toBeCalled();

    act(() => jest.advanceTimersByTime(500));
    expect(getTrigger()).not.toHaveFocus();
    expect(onBlur).toBeCalled();

    expect(onFocus.mock.calls.length).toBe(1);
    expect(onBlur.mock.calls.length).toBe(1);

    function App() {
      const options: SelectOptionsItem[] = [
        {
          label: 'foo',
          value: 1,
        },
        {
          label: 'bar',
          value: 2,
        },
        {
          label: 'bar',
          value: 3,
        },
        {
          label: 'bar',
          value: 4,
        },
        {
          label: 'bar',
          value: 5,
        },
      ];
      const controllerRef = useRef<SelectControllerRef>(null);
      return (
        <>
          <button
            onClick={() => {
              controllerRef.current?.focus();
              setTimeout(() => {
                controllerRef.current?.blur();
              }, 500);
            }}
            id="focus"
          >
            focus
          </button>
          <Select
            controllerRef={controllerRef}
            options={options}
            onFocus={onFocus}
            onBlur={onBlur}
            maxTagCount={1}
            value={[1, 2]}
            multiple
          />
        </>
      );
    }
    function getTrigger() {
      return $('.t-select-tab-trigger');
    }
  });

  test('icon', () => {
    expect(
      render(<Select options={[]} icon="ic" />).container.querySelector(
        '.t-select__suffix',
      ),
    ).toMatchSnapshot();
  });

  test('loading', () => {
    expect(
      render(<Select options={[]} loading />).container.querySelector(
        '.t-select__suffix',
      ),
    ).toMatchSnapshot();
  });

  test('status', () => {
    expect(
      render(<Select status="error" options={[]} />).container.firstChild,
    ).toHaveClass(' t-select--error');

    expect(
      render(<Select status="warning" options={[]} />).container.firstChild,
    ).toHaveClass(' t-select--warning');
  });

  test('missing options', () => {
    // 使用jest.spyOn监听prop-types警告
    const consoleErrorSpy = jest.spyOn(console, 'error');
    consoleErrorSpy.mockImplementation(emptyFn);

    render(<Select options={undefined as any} />);

    // 验证控制台是否打印了prop-types警告
    expect(consoleErrorSpy.mock.calls[0]![2]).toBe(
      'The prop `options` is marked as required in `Select`, but its value is `undefined`.',
    );
    // 恢复console.error函数
    consoleErrorSpy.mockRestore();
  });
});
