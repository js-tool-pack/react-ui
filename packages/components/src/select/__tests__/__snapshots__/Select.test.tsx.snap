// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select basic 1`] = `
<label
  class="t-input-skin t-select t-input-popover t--size-m"
>
  <input
    class="t-input-popover-tab-trigger"
    tabindex="0"
  />
  <div
    class="t-select-selection t--size-m"
  >
    <div
      class="t-select-selection__selected"
    >
      <div
        class="t-select-selection__placeholder"
      >
        Please select
      </div>
    </div>
    <div
      class="t-select-selection__suffix"
    >
      <i
        class="t-icon t-select-selection__icon"
        role="img"
      >
        <svg
          viewBox="0 0 1024 1024"
        >
          <path
            d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
            fill="currentColor"
          />
        </svg>
      </i>
    </div>
  </div>
</label>
`;

exports[`Select clearable|onClear 1`] = `
<label
  class="t-input-skin t-select t-select--selected t-select--clearable t-input-popover t--size-m"
>
  <input
    class="t-input-popover-tab-trigger"
    tabindex="0"
  />
  <div
    class="t-select-selection t--size-m"
  >
    <div
      class="t-select-selection__selected"
    >
      <div
        class="t-select-selection__label"
      >
        foo
      </div>
    </div>
    <div
      class="t-select-selection__suffix"
    >
      <i
        class="t-icon t-select-selection__icon"
        role="img"
      >
        <svg
          viewBox="0 0 1024 1024"
        >
          <path
            d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
            fill="currentColor"
          />
        </svg>
      </i>
      <i
        class="t-icon t-select-selection__clear"
        role="img"
      >
        <svg
          viewBox="0 0 1024 1024"
        >
          <path
            d="M512 14.8973037C238.59351703 14.8973037 14.8973037 238.59351703 14.8973037 512s223.69621333 497.1026963 497.1026963 497.1026963 497.1026963-223.69621333 497.1026963-497.1026963S785.40648297 14.8973037 512 14.8973037zM719.12612347 661.13080889c16.57008987 16.57008987 16.57008987 41.42522469 0 57.99531458s-41.42522469 16.57008987-57.99531458 0L512 569.99531457l-149.13080889 149.1308089c-16.57008987 16.57008987-41.42522469 16.57008987-57.99531458 0s-16.57008987-41.42522469 0-57.99531458L454.00468543 512 304.87387653 362.86919111C288.30378667 346.29910124 288.30378667 321.44396642 304.87387653 304.87387653s41.42522469-16.57008987 57.99531458 0L512 454.00468543l149.13080889-149.1308089c16.57008987-16.57008987 41.42522469-16.57008987 57.99531458 0s16.57008987 41.42522469 0 57.99531458L569.99531457 512 719.12612347 661.13080889z"
            fill="currentColor"
          />
        </svg>
      </i>
    </div>
  </div>
</label>
`;

exports[`Select disabled 1`] = `
<label
  class="t-input-skin t-select t-input-popover t--size-m t-input-skin--disabled"
>
  <input
    class="t-input-popover-tab-trigger"
    tabindex="-1"
  />
  <div
    class="t-select-selection t--size-m"
  >
    <div
      class="t-select-selection__selected"
    >
      <div
        class="t-select-selection__placeholder"
      >
        Please select
      </div>
    </div>
    <div
      class="t-select-selection__suffix"
    >
      <i
        class="t-icon t-select-selection__icon"
        role="img"
      >
        <svg
          viewBox="0 0 1024 1024"
        >
          <path
            d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
            fill="currentColor"
          />
        </svg>
      </i>
    </div>
  </div>
</label>
`;

exports[`Select group 1`] = `
<body>
  <div>
    <label
      class="t-input-skin t-select t-input-popover t--size-m t-input-skin--active"
    >
      <input
        class="t-input-popover-tab-trigger"
        tabindex="-1"
      />
      <div
        class="t-select-selection t--size-m"
      >
        <div
          class="t-select-selection__selected"
        >
          <div
            class="t-select-selection__placeholder"
          >
            Please select
          </div>
        </div>
        <div
          class="t-select-selection__suffix"
        >
          <i
            class="t-icon t-select-selection__icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
                fill="currentColor"
              />
            </svg>
          </i>
        </div>
      </div>
    </label>
  </div>
  <div
    class="t-word-balloon t-select__dropdown t-word-balloon--bottom t-word-balloon--no-arrow t-select__dropdown-enter-active t-select__dropdown-enter-to"
    style="width: 0px; top: 10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      <div
        class="t-select-menu__body"
      >
        <ul
          class="t-select-menu__options"
        >
          <li
            class="t-select-menu__group test"
          >
            <div
              class="t-option t-select-option t-select-menu__group-title t--size-m t-option--readonly"
              role="heading"
            >
              <div
                class="t-option__label"
              >
                1
              </div>
            </div>
            <ul
              class="t-select-menu__group-body"
            >
              <li
                class="t-option t-select-option t-select-option--picked t--size-m"
                role="option"
              >
                <div
                  class="t-option__label"
                >
                  foo
                </div>
              </li>
              <li
                class="t-option t-select-option t--size-m"
                role="option"
              >
                <div
                  class="t-option__label"
                >
                  bar
                </div>
              </li>
            </ul>
          </li>
          <li
            class="t-option t-select-option t--size-m"
            role="option"
          >
            <div
              class="t-option__label"
            >
              foo
            </div>
          </li>
          <li
            class="t-option t-select-option t--size-m"
            role="option"
          >
            <div
              class="t-option__label"
            >
              bar
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;

exports[`Select icon 1`] = `null`;

exports[`Select loading 1`] = `null`;

exports[`Select maxTagCount 1`] = `
<body>
  <div>
    <label
      class="t-input-skin t-select t-select--selected t-input-popover t--size-m"
    >
      <input
        class="t-input-popover-tab-trigger"
        tabindex="0"
      />
      <div
        class="t-select-selection t--size-m"
      >
        <div
          class="t-select-selection__selected"
        >
          <div
            class="t-tag t-select-tags__tag t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              foo
            </span>
            <button
              class="t-button t-tag__close t-button--type-info t-button--icon-l t-button--plain-text t--size-sm"
              tabindex="-1"
            >
              <span>
                <i
                  class="t-icon"
                  role="img"
                >
                  <svg
                    viewBox="0 0 1024 1024"
                  >
                    <path
                      d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
                      fill="currentColor"
                    />
                  </svg>
                </i>
              </span>
            </button>
          </div>
          <div
            class="t-tag t-select-tags__tag t-select-tags__count t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              + 
              1
            </span>
          </div>
        </div>
        <div
          class="t-select-selection__suffix"
        >
          <i
            class="t-icon t-select-selection__icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
                fill="currentColor"
              />
            </svg>
          </i>
        </div>
      </div>
    </label>
  </div>
</body>
`;

exports[`Select maxTagCount 2`] = `
<body>
  <div>
    <label
      class="t-input-skin t-select t-select--selected t-input-popover t--size-m"
    >
      <input
        class="t-input-popover-tab-trigger"
        tabindex="0"
      />
      <div
        class="t-select-selection t--size-m"
      >
        <div
          class="t-select-selection__selected"
        >
          <div
            class="t-tag t-select-tags__tag t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              foo
            </span>
            <button
              class="t-button t-tag__close t-button--type-info t-button--icon-l t-button--plain-text t--size-sm"
              tabindex="-1"
            >
              <span>
                <i
                  class="t-icon"
                  role="img"
                >
                  <svg
                    viewBox="0 0 1024 1024"
                  >
                    <path
                      d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
                      fill="currentColor"
                    />
                  </svg>
                </i>
              </span>
            </button>
          </div>
          <div
            class="t-tag t-select-tags__tag t-select-tags__count t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              + 
              1
            </span>
          </div>
        </div>
        <div
          class="t-select-selection__suffix"
        >
          <i
            class="t-icon t-select-selection__icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
                fill="currentColor"
              />
            </svg>
          </i>
        </div>
      </div>
    </label>
  </div>
  <div
    class="t-word-balloon t-popover t-word-balloon--top"
    style="top: -10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      <div
        class="t-select-tags__pop-tags"
      >
        <div
          class="t-tag t-select-tags__tag t--size-m t-tag--info"
        >
          <span
            class="t-tag__content"
          >
            bar
          </span>
          <button
            class="t-button t-tag__close t-button--type-info t-button--icon-l t-button--plain-text t--size-sm"
            tabindex="-1"
          >
            <span>
              <i
                class="t-icon"
                role="img"
              >
                <svg
                  viewBox="0 0 1024 1024"
                >
                  <path
                    d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
                    fill="currentColor"
                  />
                </svg>
              </i>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="t-word-balloon__arrow"
    >
      <svg
        viewBox="-8 -5.5 16 5.515"
      >
        <path
          d="M8-5.5A4 4 180 005.1716-4.3284L1.4142-.5711A2 2 180 01-1.4142-.5711L-5.1716-4.3284A4 4 180 00-8-5.5Z"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Select multiple 1`] = `
<body>
  <div>
    <label
      class="t-input-skin t-select t-select--selected t-input-popover t--size-m t-input-skin--active"
    >
      <input
        class="t-input-popover-tab-trigger"
        tabindex="-1"
      />
      <div
        class="t-select-selection t--size-m"
      >
        <div
          class="t-select-selection__selected"
        >
          <div
            class="t-tag t-select-tags__tag t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              foo
            </span>
            <button
              class="t-button t-tag__close t-button--type-info t-button--icon-l t-button--plain-text t--size-sm"
              tabindex="-1"
            >
              <span>
                <i
                  class="t-icon"
                  role="img"
                >
                  <svg
                    viewBox="0 0 1024 1024"
                  >
                    <path
                      d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
                      fill="currentColor"
                    />
                  </svg>
                </i>
              </span>
            </button>
          </div>
          <div
            class="t-tag t-select-tags__tag t--size-m t-tag--info"
          >
            <span
              class="t-tag__content"
            >
              bar
            </span>
            <button
              class="t-button t-tag__close t-button--type-info t-button--icon-l t-button--plain-text t--size-sm"
              tabindex="-1"
            >
              <span>
                <i
                  class="t-icon"
                  role="img"
                >
                  <svg
                    viewBox="0 0 1024 1024"
                  >
                    <path
                      d="M574.56865938 513.72328935l366.30945353-364.40941657c17.3212673-17.23289347 17.4096411-45.15901828 0.17674762-62.48028556-17.27708037-17.36545419-45.15901828-17.4096411-62.48028556-0.17674764l-366.39782734 364.45360348-363.21637009-364.32104275C131.77167097 69.55650684 103.80135925 69.46813303 86.48009197 86.70102651 69.20301159 103.93391997 69.15882467 131.90423169 86.39171815 149.22549897l363.12799629 364.23266894L83.38700853 877.64664993c-17.3212673 17.23289347-17.4096411 45.15901828-0.17674763 62.48028558 8.61644674 8.66063363 19.97248167 13.03513737 31.32851659 13.03513736 11.26766112 0 22.53532223-4.28612991 31.15176897-12.85838974l366.22107973-364.27685584 366.53038806 367.63506071c8.61644674 8.66063363 19.92829476 12.99095045 31.2843297 12.99095047 11.31184801 0 22.57950913-4.33031682 31.19595588-12.90257666 17.27708037-17.23289347 17.3212673-45.15901828 0.0883738-62.48028554L574.56865938 513.72328935z"
                      fill="currentColor"
                    />
                  </svg>
                </i>
              </span>
            </button>
          </div>
        </div>
        <div
          class="t-select-selection__suffix"
        >
          <i
            class="t-icon t-select-selection__icon"
            role="img"
          >
            <svg
              viewBox="0 0 1024 1024"
            >
              <path
                d="M512.22982846 638.25693048L92.98520897 219.20383471 11.74100274 300.4097362l500.52713049 500.25899727 500.02916876-500.25899727-81.28251096-81.20590149z"
                fill="currentColor"
              />
            </svg>
          </i>
        </div>
      </div>
    </label>
  </div>
  <div
    class="t-word-balloon t-select__dropdown t-word-balloon--bottom t-word-balloon--no-arrow t-select__dropdown-enter-active t-select__dropdown-enter-to"
    style="width: 0px; top: 10px; left: 0px;"
  >
    <div
      class="t-word-balloon__content"
    >
      <div
        class="t-select-menu__body"
      >
        <ul
          class="t-select-menu__options"
        >
          <li
            class="t-option t-select-option t-select-option--selected t-select-option--picked t--size-m"
            role="option"
          >
            <div
              class="t-option__label"
            >
              foo
            </div>
            <div
              class="t-option__extra"
            >
              <i
                class="t-icon t-select-option__selected-icon"
                role="img"
              >
                <svg
                  viewBox="0 0 1024 1024"
                >
                  <path
                    d="M76.35918602 484.02337646l-71.37237435 71.50703811 331.54487886 330.63589206 682.31316641-674.94026626-71.03571155-71.77636833-610.97445834 604.37587985z"
                    fill="currentColor"
                  />
                </svg>
              </i>
            </div>
          </li>
          <li
            class="t-option t-select-option t-select-option--selected t--size-m"
            role="option"
          >
            <div
              class="t-option__label"
            >
              bar
            </div>
            <div
              class="t-option__extra"
            >
              <i
                class="t-icon t-select-option__selected-icon"
                role="img"
              >
                <svg
                  viewBox="0 0 1024 1024"
                >
                  <path
                    d="M76.35918602 484.02337646l-71.37237435 71.50703811 331.54487886 330.63589206 682.31316641-674.94026626-71.03571155-71.77636833-610.97445834 604.37587985z"
                    fill="currentColor"
                  />
                </svg>
              </i>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
